design/validation-spi.md

```diff
@@ -0,0 +1,293 @@
+    <groupId>org.hibernate</groupId>
+    <artifactId>hibernate-validator</artifactId>
+    <version>6.0.12.Final</version>
+</dependency>
+```
+- No direct support for validation (Constraint) lookups by name. The user-profile support needs a way to lookup validations by name.
+- No easy way to dynamically configure constraints programatically. The user-profile support needs a way to configure validation dynamically.
+- Definition of [custom validations is quite involved](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#validator-customconstraints) - you need 2 classes (annotation + constraint) + error message + META-INF service manifest.
+
+All this was considered and evaluated in a small prototype, but then it was concluded that it makes more sense to go 
+with a custom implementation that provides a narrow API. Additionally custom extensions should be easier and provide 
+a better integration with the existing Keycloak infrastructure. Further more dynamic validation logic lookups 
+and configuration should be possible and easy to do. 
+
+Note that it is still possible to provide a validation/validator implementation that leverages bean validation. 
+Also, bean-validation still works quite well for declarative validation in JAX-RS endpoints.
+
+### Fully fledged Validation SPI
+
+An earlier attempt at a unified validation SPI can be found here [KEYCLOAK-2045 Add support for flexible Validation](https://github.com/keycloak/keycloak/pull/7324) and was discussed on the [keycloak-dev mailing list](https://groups.google.com/g/keycloak-dev/c/-XjQu7rn56s).
+
+After some discussion, it became apparent that this proposal provides a very comprehensive validation framework that 
+offers numerous functionalities that go beyond the requirements of user profile validation and other validations in Keycloak.
+
+However, the API was very complex and not easy to understand. Therefore, it was agreed that @thomasdarimont would revise 
+the proposal again and present it in a smaller form. This has resulted in the new "Simple Validation SPI" proposal.
+
+## Proposed Validation SPI
+
+### PoC for new Validator SPI
+
+The Proof of concept implementation can be found here in the PR [KEYCLOAK-2045 Simple Validation API #7887](https://github.com/keycloak/keycloak/pull/7887).
+
+Some usage examples can be found here in the [ValidatorTest](https://github.com/thomasdarimont/keycloak/blob/issue/KEYCLOAK-2045-Simple-Validation-SPI/server-spi-private/src/test/java/org/keycloak/validate/ValidatorTest.java).
+
+### Core API
+The proposal focus around the following types:
+1. `Validator`: `Provider` interface for validation mechanics.
+1. `ValidationContext`: Holds state of a sequence of validations.
+1. `ValidationResult`: Denotes the outcome of a validation.
+1. `ValidatorConfig`: Denotes the configuration for a Validator. A typed wrapper around a `Map<String,Object>`.
+1. `ValidationError`: Represents an error found during validation.
```

Comment:
I `ValidationError` any different ot `ValidationResult`?

Reply:
A `ValidationError` denotes a concrete error found during validation that was reported by a `Validator`. `ValidationResult` contains a collection of all the errors found during validation.